lxd/storage/drivers/driver_pure_util.go

```diff
@@ -12,3 +49,186 @@ func newPureClient(driver *pure) *pureClient {
 		driver: driver,
 	}
 }
+
+// createBodyReader creates a reader for the given request body contents.
+func (p *pureClient) createBodyReader(contents map[string]any) (io.Reader, error) {
+	body := &bytes.Buffer{}
+
+	err := json.NewEncoder(body).Encode(contents)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to write request body: %w", err)
+	}
+
+	return body, nil
+}
+
+// request issues a HTTP request against the PureStorage gateway.
+func (p *pureClient) request(method string, path string, reqBody io.Reader, reqHeaders map[string]string, respBody any, respHeaders map[string]string) error {
+	var url string
+
+	// Construct the request URL.
+	if strings.HasPrefix(path, "/api") {
+		// If the provided path starts with "/api", simply append it to the gateway URL.
+		url = fmt.Sprintf("%s%s", p.driver.config["pure.gateway"], path)
+	} else {
+		// Otherwise, prefix the path with "/api/<api_version>" and then append it to the gateway URL.
+		// If API version is not known yet, retrieve and cache it first.
+		if p.driver.apiVersion == "" {
+			apiVersions, err := p.getAPIVersions()
+			if err != nil {
+				return fmt.Errorf("Failed to retrieve supported PureStorage API versions: %w", err)
+			}
+
+			// Use the latest available API version.
+			p.driver.apiVersion = apiVersions[len(apiVersions)-1]
+		}
+
+		url = fmt.Sprintf("%s/api/%s%s", p.driver.config["pure.gateway"], p.driver.apiVersion, path)
+	}
+
+	req, err := http.NewRequest(method, url, reqBody)
+	if err != nil {
+		return fmt.Errorf("Failed to create request: %w", err)
+	}
+
+	// Set custom request headers.
+	for k, v := range reqHeaders {
+		req.Header.Add(k, v)
+	}
+
+	req.Header.Add("Accept", "application/json")
+	if reqBody != nil {
+		req.Header.Add("Content-Type", "application/json")
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: shared.IsFalse(p.driver.config["pure.gateway.verify"]),
+			},
+		},
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("Failed to send request: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	// Wrap unauthorized requests into an API status error.
+	if resp.StatusCode == http.StatusUnauthorized {
+		return api.StatusErrorf(http.StatusUnauthorized, "Unauthorized request")
```

Comment:
Are you masking the response body on purpose (it might contain useful info right)?